migration/migration_test.go

```diff
@@ -0,0 +1,332 @@
+	"github.com/stretchr/testify/require"
+	mongods "github.com/textileio/go-ds-mongo"
+	"github.com/textileio/powergate/tests"
+
+	logger "github.com/ipfs/go-log/v2"
+)
+
+func TestEmptyDatastore(t *testing.T) {
+	t.Parallel()
+
+	ms := map[int]Migration{
+		1: {
+			UseTxn: true,
+			Run: func(_ datastoreReaderWriter) error {
+				return fmt.Errorf("This migration shouldn't be run on empty database")
+			},
+		},
+	}
+	m := newMigrator(ms, true)
+
+	v, empty, err := m.getCurrentVersion()
+	require.NoError(t, err)
+	require.Equal(t, 0, v)
+	require.True(t, empty)
+
+	err = m.Ensure()
+	require.NoError(t, err)
+
+	v, _, err = m.getCurrentVersion()
+	require.NoError(t, err)
+	require.Equal(t, 1, v)
+}
+
+func TestNonEmptyDatastore(t *testing.T) {
+	t.Parallel()
+
+	ms := map[int]Migration{
+		1: {
+			UseTxn: true,
+			Run: func(_ datastoreReaderWriter) error {
+				return nil
+			},
+		},
+	}
+	m := newMigrator(ms, false)
+
+	v, empty, err := m.getCurrentVersion()
+	require.NoError(t, err)
+	require.Equal(t, 0, v)
+	require.False(t, empty)
+
+	err = m.Ensure()
+	require.NoError(t, err)
+
+	v, _, err = m.getCurrentVersion()
+	require.NoError(t, err)
+	require.Equal(t, 1, v)
+}
+
+func TestForwardOnly(t *testing.T) {
+	t.Parallel()
+
+	ms := map[int]Migration{
+		1: {
+			UseTxn: true,
+			Run: func(_ datastoreReaderWriter) error {
+				return nil
+			},
+		},
+	}
+
+	m := newMigrator(ms, false)
+	err := m.bootstrapEmptyDatastore(10)
+	require.NoError(t, err)
+
+	v, _, err := m.getCurrentVersion()
+	require.NoError(t, err)
+	require.Equal(t, 10, v)
+
+	err = m.Ensure()
+	require.Error(t, err)
+}
+
+func TestNoop(t *testing.T) {
+	t.Parallel()
+
+	ms := map[int]Migration{
+		1: {
+			UseTxn: true,
+			Run: func(_ datastoreReaderWriter) error {
+				return nil
+			},
+		},
+	}
+	m := newMigrator(ms, false)
+
+	err := m.Ensure()
+	require.NoError(t, err)
+
+	err = m.Ensure()
+	require.NoError(t, err)
+}
+
+func TestFailingMigration(t *testing.T) {
+	ms := map[int]Migration{
+		1: {
+			UseTxn: true,
+			Run: func(_ datastoreReaderWriter) error {
+				return fmt.Errorf("I failed")
+			},
+		},
+	}
+	m := newMigrator(ms, false)
+
+	v, _, err := m.getCurrentVersion()
+	require.NoError(t, err)
+	require.Equal(t, 0, v)
+
+	err = m.Ensure()
+	require.Error(t, err)
+
+	v, _, err = m.getCurrentVersion()
+	require.NoError(t, err)
+	require.Equal(t, 0, v)
+}
+
+func TestRealDataBadger(t *testing.T) {
+	logger.SetDebugLogging()
+	_ = logger.SetLogLevel("badger", "error")
+	tmpDir := t.TempDir()
+	err := copyDir("testdata/badgerdump", tmpDir+"/badgerdump")
+	require.NoError(t, err)
+
+	opts := &badger.DefaultOptions
+	ds, err := badger.NewDatastore(tmpDir+"/badgerdump", opts)
+	require.NoError(t, err)
+	defer func() { require.NoError(t, ds.Close()) }()
+
+	migrations := map[int]Migration{
+		1: V1MultitenancyMigration,
+	}
+	m := New(ds, migrations)
+	err = m.Ensure()
+	require.NoError(t, err)
+}
+
+func TestRealDataRemoteMongo(t *testing.T) {
+	t.SkipNow()
```

Comment:
This test case was mostly to test the migration directly in copied datastores of many live hosted powergates.
Disabled the test since doesn't make sense for CI.